### PR TITLE
[BI-677] - PUT traits

### PR DESCRIPTION
**Pull request by ctucker3**
Add PUT traits to the backend. PUT traits accepts a list of traits. Also pushes the changes to the connected brapi server. 

Added a positive test for PUT traits. Needs a couple more negative traits. 

See other PRs:
https://github.com/Breeding-Insight/bi-web/pull/72
https://github.com/Breeding-Insight/sgn/pull/7

### CODE REVIEW

**Review by timparsons — APPROVED** [review]
One small comment that doesn't need to hold up approval.  Looks good!

File: src/main/java/org/breedinginsight/services/TraitService.java [hunk 1]
```diff
@@ -287,4 +232,163 @@ public void preprocessTraits(List<Trait> traits) {
...
+                    }
+
+                    // Jump to scale
+                    ScaleEntity existingScaleEntity = scaleDAO.fetchOneById(existingTraitEntity.getScaleId());
+                    existingScaleEntity.setScaleName(updatedTrait.getScale().getScaleName());
+                    existingScaleEntity.setDataType(updatedTrait.getScale().getDataType());
+                    existingScaleEntity.setUpdatedBy(user.getId());
+                    scaleDAO.update(existingScaleEntity);
+
+                    // Update trait
```
Comment (nickpalladino) [inline src/main/java/org/breedinginsight/services/TraitService.java:377]: We also need to set active right?

File: src/test/java/org/breedinginsight/api/v1/controller/TraitControllerIntegrationTest.java [hunk 2]
```diff
@@ -914,6 +914,67 @@ public void postTraitComputation() {
...
+        List<String> jsonAlternativeAbbreviations = new ArrayList<>();
+        trait.get("abbreviations").getAsJsonArray().iterator().forEachRemaining(element -> jsonAlternativeAbbreviations.add(element.getAsString()));
+        List<String> traitAbbreviations = Arrays.asList(updateTrait.getAbbreviations());
+        Collections.sort(jsonAlternativeAbbreviations);
+        Collections.sort(traitAbbreviations);
+        assertLinesMatch(traitAbbreviations, jsonAlternativeAbbreviations, "Alternative abbreviations don't match");
+
+    }
+
+    //TODO: Test for passing a variable in with an id that doesn't exist
```
[the remark this answers is not in the document]
Comment (ctucker3) [reply src/test/java/org/breedinginsight/api/v1/controller/TraitControllerIntegrationTest.java]: Dang! I would have gotten away with it too if it weren't for you pesky kids! 

I implemented these tests and pushed them up.

File: src/test/java/org/breedinginsight/api/v1/controller/TraitControllerIntegrationTest.java [hunk 3]
```diff
@@ -914,6 +914,67 @@ public void postTraitComputation() {
...
+        assertEquals(updateTrait.getProgramObservationLevel().getName(),
+                trait.get("programObservationLevel").getAsJsonObject().get("name").getAsString(), "wrong observation level");
+        assertEquals(updateTrait.getScale().getScaleName(),
+                trait.get("scale").getAsJsonObject().get("scaleName").getAsString(), "wrong scale name");
+        assertEquals(updateTrait.getScale().getDataType().toString(),
+                trait.get("scale").getAsJsonObject().get("dataType").getAsString(), "wrong scale data type");
+        assertEquals(updateTrait.getMethod().getDescription(),
+                trait.get("method").getAsJsonObject().get("description").getAsString(), "wrong method description");
+        //TODO: Reintroduce variables stored in brapi-server. The brapi test server is broken right now. See https://github.com/plantbreeding/API/issues/455
+        //assertEquals(updateTrait.getMethod().getMethodClass(), trait.get("method").getAsJsonObject().get("methodClass").getAsString(), "wrong method class");
```
[the remark this answers is not in the document]
Comment (ctucker3) [reply src/test/java/org/breedinginsight/api/v1/controller/TraitControllerIntegrationTest.java:965]: Created a new card:

https://breedinginsight.atlassian.net/browse/BI-731

File: src/main/java/org/breedinginsight/services/TraitService.java (hunk 1 above)
Comment (ctucker3) [reply src/main/java/org/breedinginsight/services/TraitService.java:377]: Wouldn't that be in a separate archive endpoint?

File: src/main/java/org/breedinginsight/api/v1/controller/TraitController.java [hunk 4]
```diff
@@ -142,4 +142,29 @@ public TraitController(TraitService traitService, SecurityService securityServic
...
+    @Put("/programs/{programId}/traits")
+    @Produces(MediaType.APPLICATION_JSON)
+    @ProgramSecured(roles = {ProgramSecuredRole.BREEDER})
+    public HttpResponse<Response<DataResponse<Trait>>> updateTraits(@PathVariable UUID programId, @Body @Valid List<Trait> traits) {
+        AuthenticatedUser actingUser = securityService.getUser();
+        try {
+            List<Trait> updatedTraits = traitService.updateTraits(programId, traits, actingUser);
+            List<Status> metadataStatus = new ArrayList<>();
+            // Users query successfully
+            metadataStatus.add(new Status(StatusCode.INFO, "Successful Query"));
```
[the remark this answers is not in the document]
Comment (ctucker3) [reply src/main/java/org/breedinginsight/api/v1/controller/TraitController.java]: Updated to `Successful Creation`

File: src/main/java/org/breedinginsight/services/TraitService.java (hunk 1 above)
Comment (nickpalladino) [reply src/main/java/org/breedinginsight/services/TraitService.java:377]: based on later discussions, yes

**Review by nickpalladino — APPROVED** [review]

**Pull request merged**